File: sources/webcore/opt/webcore/libexec/account.py
#!/usr/bin/env python
# --------------------------------
# Copyright (c) 2011 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
# ---------------------------------

import sys, os, logging, json

import bottle
from bottle import route, get, put, delete, request, HTTPError, post, response

## Canopsis
from caccount import caccount
from cstorage import cstorage
from cstorage import get_storage
from crecord import crecord
try:
	from cgroup import cgroup
except:
	pass

#import protection function
from libexec.auth import check_auth, get_account
#

logger = logging.getLogger('Account')

#########################################################################

#### GET Me
@get('/account/me',apply=[check_auth])
def account_get_me():
	namespace = 'object'
	ctype= 'account'
	
	#get the session (security)
	account = get_account()

	storage = get_storage(namespace=namespace)

	#try:
	logger.debug(" + Try to get '%s' ... " % account._id)
	record = storage.get(account._id, account=account)

	logger.debug("   + Result: '%s'" % record)
	#except Exception, err:
	#	self.logger.error("Exception !\nReason: %s" % err)
	#	return HTTPError(404, _id+" Not Found")

	if record:
		data = record.dump(json=True)
		data['id'] = data['_id']
		output = [data]

	output={'total': 1, 'success': True, 'data': output}

	logger.debug(" + Output: "+str(output))
	
	logger.debug('Response status: %s' % response.status)
	
	return output

#### POST setConfig
@post('/account/setConfig/:_id',apply=[check_auth])
def account_setConfig(_id):
	account = get_account()
	storage = get_storage(namespace='object')
	
	value = request.params.get('value', default=None)
	
	logger.debug(" + setConfig '%s' => '%s'" % (_id, value))
	
	if value:
		account.data[_id] = value
		storage.put(account, account=account)
		output={'total': 0, 'success': True, 'data': []}
	else:
		output={'total': 0, 'success': False, 'data': []}
	
	return output

@get('/account/getAuthKey/:dest_account',apply=[check_auth])
def account_getAuthKey(dest_account):
	if not dest_account:
		return HTTPError(404, 'No account specified')
	
	#------------------------get accounts----------------------
	account = get_account()
	storage = get_storage(namespace='object',account=account)
	
	_id = 'account.%s' % dest_account
	
	try:
		aim_account = caccount(storage.get(_id,account=account))
		
		return {'total':1,'success':True,'data':{'authkey':aim_account.get_authkey()}}
	except Exception,err:
		logger.debug('Error while fetching account : %s' % err)
		return {'total':0,'success':False,'data':{'output':str(err)}}
		
	

@get('/account/getNewAuthKey/:dest_account',apply=[check_auth])
def account_newAuthKey(dest_account):
	if not dest_account:
		return HTTPError(404, 'No account specified')
	
	#------------------------get accounts----------------------
	account = get_account()
	storage = get_storage(namespace='object',account=account)
	
	_id = 'account.%s' % dest_account
	
	try:
		aim_account = caccount(storage.get(_id,account=account))
	except:
		logger.debug('aimed account not found')
		return HTTPError(404, 'Wrong account name or no enough rights')
	
	#---------------------generate new key-------------------
	logger.debug('Change AuthKey for : %s' % aim_account.user)
	
	try:
		aim_account.generate_new_authkey()
		storage.put(aim_account)
		logger.debug('New auth key is : %s' % aim_account.get_authkey())
		return {'total': 0, 'success': True, 'data': {'authkey': aim_account.get_authkey(),'account':aim_account.user}}
	except Exception,err:
		logger.error('Error while updating auth key : %s' % err)
		return {'total': 0, 'success': False, 'data': {}}
	
#### GET
@get('/account/:_id',apply=[check_auth])
@get('/account/',apply=[check_auth])
def account_get(_id=None):
	namespace = 'object'
	ctype= 'account'
	
	#get the session (security)
	account = get_account()

	limit = int(request.params.get('limit', default=20))
	page =  int(request.params.get('page', default=0))
	start =  int(request.params.get('start', default=0))
	groups = request.params.get('groups', default=None)
	search = request.params.get('search', default=None)

	logger.debug("GET:")
	logger.debug(" + User: "+str(account.user))
	logger.debug(" + Group(s): "+str(account.groups))
	logger.debug(" + namespace: "+str(namespace))
	logger.debug(" + Ctype: "+str(ctype))
	logger.debug(" + _id: "+str(_id))
	logger.debug(" + Limit: "+str(limit))
	logger.debug(" + Page: "+str(page))
	logger.debug(" + Start: "+str(start))
	logger.debug(" + Groups: "+str(groups))
	logger.debug(" + Search: "+str(search))

	storage = get_storage(namespace=namespace)

	total = 0
	mfilter = {}
	if ctype:
		mfilter = {'crecord_type': ctype}
	if _id:	
		try:
			records = [ storage.get(_id, account=account) ]
			total = 1
		except Exception, err:
			self.logger.error("Exception !\nReason: %s" % err)
			return HTTPError(404, _id+" Not Found")
		
	else:
		records =  storage.find(mfilter, limit=limit, offset=start, account=account)
		total =	   storage.count(mfilter, account=account)

	output = []
	for record in records:
		if record:
			data = record.dump(json=True)
			data['id'] = data['_id']
			output.append(data)

	output={'total': total, 'success': True, 'data': output}

	#logger.debug(" + Output: "+str(output))

	return output

	
#### POST
@post('/account/', apply=[check_auth])
def account_post():
	#get the session (security)
	account = get_account()
	root_account = caccount(user="root", group="root")
	
	storage = get_storage(namespace='object',account=account)

	logger.debug("POST:")

	data = request.body.readline()
	if not data:
		return HTTPError(400, "No data received")

	data = json.loads(data)

	## Clean data
	try:
		del data['_id']
		del data['id']
		del data['crecord_type']
	except:
		pass
	
	if data['user']:
		_id = "account." + str(data['user'])

		#check if already exist
		update = False
		try:
			record = storage.get(_id ,account=account)
			logger.debug('Update account %s' % _id)
			update = True
		except:
			logger.debug('Create account %s' % _id)

		#-----------------------UPDATE----------------------
		if update:
			#Get password
			passwd = str(data['passwd'])
				
			#Get group
			group = str(data['aaa_group'])
			if group:
				if group.find('group.') == -1:
					group = 'group.%s' % group
			
			#get secondary groups
			groups = data['groups']
			secondary_groups = []
			if groups:
				if not isinstance(groups,list):
					groups = [groups]
				for one_group in groups:
					if one_group.find('group.') == -1:
						one_group = 'group.%s' % one_group
						try :
							secondary_groups.append(cgroup(storage.get(one_group,account=account)))
						except Exception,err:
							logger.error('Error while searching secondary group: %s',err)
			
			#clean
			del data['passwd']
			del data['aaa_group']
			del data['groups']

			#new record
			for key in dict(data).keys():
				record.data[key] = data[key]
			update_account = caccount(record)
			
			#updating
			if passwd:
				logger.debug(' + Update password ...')
				update_account.passwd(passwd)
			if group:
				logger.debug(' + Update group ...')
				update_account.chgrp(group)
			if secondary_groups:
				logger.debug(' + Update groups ...')
				#clean
				update_account.groups = []
				update_account.add_in_groups(secondary_groups)
				logger.debug(update_account.dump())

			storage.put(update_account, account=account)
			storage.put(secondary_groups, account=account)

		else:
			#----------------------------CREATION--------------------------
			logger.debug(' + New account')
			new_account = caccount(user=data['user'], group=data['aaa_group'], lastname=data['lastname'], firstname=data['firstname'], mail=data['mail'])

			#passwd
			passwd = data['passwd']
			new_account.passwd(passwd)
			logger.debug("   + Passwd: '%s'" % passwd)

			#secondary groups
			groups = data['groups']
			secondary_groups = []
			if groups:
				if not isinstance(groups,list):
					groups = [groups]
				for one_group in groups:
					if one_group.find('group.') == -1:
						one_group = 'group.%s' % one_group
						try :
							secondary_groups.append(cgroup(storage.get(one_group,account=account)))
						except Exception,err:
							logger.error('Error while searching secondary group: %s',err)
			
			new_account.add_in_groups(secondary_groups)
			storage.put(secondary_groups)
			
			#put record
			logger.debug(' + Save new account')
			new_account.chown(new_account._id)
			storage.put(new_account, account=account)
			
			#get rootdir
			logger.debug(' + Create view directory')
			rootdir = storage.get('directory.root', account=root_account)
			
			if rootdir:
				userdir = crecord({'_id': 'directory.root.%s' % new_account.user,'id': 'directory.root.%s' % new_account.user ,'expanded':'true'}, type='view_directory', name=new_account.user)
				userdir.chown(new_account._id)
				userdir.chgrp(new_account.group)
				userdir.chmod('g-w')
				userdir.chmod('g-r')
				storage.put(userdir, account=account)
				rootdir.add_children(userdir)

				storage.put(rootdir, account=root_account)
				storage.put(userdir, account=account)
			else:
				logger.error('Impossible to get rootdir')

	else:
		logger.warning('WARNING : no user specified ...')


#### DELETE
@delete('/account/:_id',apply=[check_auth])
def account_delete(_id):
	account = get_account()
	storage = get_storage(namespace='object')

	logger.debug("DELETE:")
	logger.debug(" + _id: "+str(_id))
	try:
		storage.remove(_id, account=account)
		logger.debug('account removed')
	except:
		return HTTPError(404, _id+" Not Found")

### GROUP
@post('/account/addToGroup/:group_id/:account_id',apply=[check_auth])
def add_account_to_group(group_id=None,account_id=None):
	if not group_id or not account_id:
		return HTTPError(400, 'Bad request, must specified group and account')
	
	session_account = get_account()
	storage = get_storage(namespace='object',account=session_account)
	
	#get group && account
	if group_id.find('group.') == -1:
		group_id = 'group.%s' % group_id
		
	if account_id.find('account.') == -1:
		account_id = 'account.%s' % account_id
		
	logger.debug('Try to get %s and %s' % (account_id,group_id))
		
	try:
		account_record = storage.get(account_id,account=session_account)
		account = caccount(account_record)
		group_record = storage.get(group_id,account=session_account)
		group = cgroup(group_record)
		
	except Exception,err:
		logger.error('error while fetching %s and %s : %s' % (account_id,group_id,err))
		return HTTPError(403, 'Record not found or insufficient rights')
		
	#put in group
	group.add_accounts(account)
	
	try:
		storage.put([group,account])
	except:
		logger.error('Put group/account in db goes wrong')
		return HTTPError(500, 'Put group/account in db goes wrong')
	
	return {'total' :1, 'success' : True, 'data':[]}
		
@post('/account/removeFromGroup/:group_id/:account_id',apply=[check_auth])
def remove_account_from_group(group_id=None,account_id=None):
	if not group_id or not account_id:
		return HTTPError(400, 'Bad request, must specified group and account')
	
	session_account = get_account()
	storage = get_storage(namespace='object',account=session_account)
	
	#get group && account
	if group_id.find('group.') == -1:
		group_id = 'group.%s' % group_id
		
	if account_id.find('account.') == -1:
		account_id = 'account.%s' % account_id
		
	logger.debug('Try to get %s and %s' % (account_id,group_id))
		
	try:
		account_record = storage.get(account_id,account=session_account)
		account = caccount(account_record)
		group_record = storage.get(group_id,account=session_account)
		group = cgroup(group_record)
		
	except Exception,err:
		logger.error('error while fetching %s and %s : %s' % (account_id,group_id,err))
		return HTTPError(403, 'Record not found or insufficient rights')
		
	#remove in group
	group.remove_accounts(account)
	
	try:
		storage.put([group,account])
	except:
		logger.error('Put group/account in db goes wrong')
		return HTTPError(500, 'Put group/account in db goes wrong')
	
	return {'total' :1, 'success' : True, 'data':[]}
		
		

File: sources/webcore/var/www/canopsis/app/controller/Account.js
/*
#--------------------------------
# Copyright (c) 2011 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
# ---------------------------------
*/
Ext.define('canopsis.controller.Account', {
	extend: 'canopsis.lib.controller.cgrid',
    
	views: ['Account.Grid', 'Account.Form'],
	stores: ['Accounts'],
	models: ['Account'],

	iconCls: 'icon-crecord_type-account',
	
	logAuthor: '[controller][Account]',

	init: function() {
		log.debug('['+this.id+'] - Initialize ...');

		this.formXtype = 'AccountForm'
		this.listXtype = 'AccountGrid'

		this.modelId = 'Account'

		this.callParent(arguments);
		
		global.accountCtrl = this
	},

	getConfig: function(id, default_value){
		log.debug(' + getConfig '+id, this.logAuthor);
		if ( ! global.account[id] ){
			if( global[id] ){
				global.account[id] = global[id]
			}else{
				global.account[id] = default_value
			}
		}
		return global.account[id]
	},
	
	setConfig: function(id, value, cb){
		log.debug(' + setConfig ' + id + ' => ' + value, this.logAuthor);
		global.account[id] = value
		
		var url = '/account/setConfig/' + id
		
		if (! cb){
			cb = function(){
				log.debug(' + setConfig Ok', this.logAuthor);
			}
		}
		
		ajaxAction(url, {value: value}, cb, this, 'POST');
		
		return global.account[id]
	},
	
	setLocale: function(locale){
		var cb = function(){
			log.debug(' + setLocale Ok', this.logAuthor);
			Ext.MessageBox.show({
				title: _('Configure language'),
				msg: _("Application must be reloaded, do you want to reload now ?"),
				icon: Ext.MessageBox.WARNING,
  				buttons: Ext.Msg.OKCANCEL,
  				fn: function(btn){
					if (btn == 'ok'){
						window.location.reload()
					}
				}
			});
		}
		
		this.setConfig('locale', locale, cb)	
	},

	beforeload_EditForm: function(form){
		var pass_textfield = Ext.ComponentQuery.query("#" + form.id + " textfield[name=passwd]")[0]
		if (pass_textfield){
			pass_textfield.allowBlank = true
		}

		var user_textfield = Ext.ComponentQuery.query("#" + form.id + " textfield[name=user]")[0]
		if (user_textfield)
			user_textfield.hide()
		
		var passwd_textfield = Ext.ComponentQuery.query("#" + form.id + " textfield[name=passwd]")[0]
		if(passwd_textfield)
			passwd_textfield.allowBlank = true

	},
	
	preSave: function(record,data,form){
		//don't update password if it's empty
		if(form.editing && (record.get('passwd') == '')){
			delete record.data.passwd
		}
		return record
	},

	validateForm: function(store, data, form){
		var already_exist = false;

		// in creation mode
		if (!form.editing) {
			store.findBy(
				function(record, id){
					if(record.get('user') == data['user']){
						log.debug('User already exist', this.logAuthor + '[validateForm]');
						already_exist = true;  // a record with this data exists
					}
				}
			);
		}

		if (already_exist){
			global.notify.notify(data['user'] + ' already exist','you can\'t add the same user twice','error')
			return false
		}else{
			return true
		}	
	},
	
	//check if user have right on this record
	check_record_right : function(record,option){
		var user = global.account.user
		var groups = global.account.groups

		//root can do everything
		if(user == 'root'){
			return true
		}
		
		if((option == 'r') || (option == 'w')){
			if ((user == record.get('aaa_owner')) && (record.data.aaa_access_owner.indexOf(option) > -1)){
				//log.debug('owner')
				return true
			} else if((groups.indexOf(record.get('aaa_group')) != -1) && (record.data.aaa_access_group.indexOf(option) > -1)){
				//log.debug('group')
				return true
			} else {
				//log.debug('nothing')
				return false
			}
		} else {
			log.error(_('Incorrect right option given'),this.logAuthor)
		}
	},
	
	//check if user have right on this obj
	check_right: function(obj,option){
		var user = global.account.user
		var groups = global.account.groups

		//root can do everything
		if(user == 'root'){
			return true
		}
		
		if((option == 'r') || (option == 'w')){
			if ((user == obj.aaa_owner) && (obj.aaa_access_owner.indexOf(option) > -1)){
				//log.debug('owner')
				return true
			} else if((groups.indexOf(obj.aaa_group) != -1) && (obj.aaa_access_group.indexOf(option) > -1)){
				//log.debug('group')
				return true
			} else {
				//log.debug('nothing')
				return false
			}
		} else {
			log.error(_('Incorrect right option given'),this.logAuthor)
		}
		
	},
	
	//if callback_func != null and ajax success -> callback is call in passed scope with
	//new key as argument
	new_authkey : function(account,callback_func,scope){
		if(account){
			//------------------------------ajax request----------------------
			log.debug('Ask webserver for new authentification key',this.logAuthor)
			Ext.Ajax.request({
				url: '/account/getNewAuthKey/' + account,
				method: 'GET',
				scope: scope,
				success: function(response){
					var object_response = Ext.decode(response.responseText)
					if(object_response.success == true){
						global.notify.notify(_('Success'),_('Your authentification key is updated'))
						var authkey = object_response.data.authkey
						global.account.authkey = authkey
						if(callback_func)
							callback_func.call(this,authkey)
					}else{
						log.error('Ajax output incorrect',this.logAuthor)
					}
				},
				failure : function(response){
					global.notify.notify(_('Error'),_('An error have occured during the updating process'),'error')
					log.error('Error while fetching new Authkey',this.logAuthor)
				}
			})
		}else{
			log.error('No account provided for Authkey')
		}
	},

	get_authkey : function(account,callback_func,scope){
		log.debug('Ask webserver for authentification key',this.logAuthor)
		Ext.Ajax.request({
			url: '/account/getAuthKey/' + account,
			method: 'GET',
			scope: scope,
			success: function(response){
				var object_response = Ext.decode(response.responseText)
				if(object_response.success == true){
					var authkey = object_response.data.authkey
					if(callback_func)
						callback_func.call(this,authkey)
				}else{
					log.error('Ajax output incorrect',this.logAuthor)
				}
			},
			failure : function(response){
				global.notify.notify(_('Error'),_('An error have occured during the process'),'error')
				log.error('Error while fetching new Authkey',this.logAuthor)
			}
		})
	}	
});
